phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -0,0 +1,269 @@
+/*                     __                                                   *\
+**     ________ ___   / /  ___      __ ____  PhantomJS support for Scala.js **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL       **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    https://www.scala-js.org/      **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                                   **
+**                          |/____/                                         **
+\*                                                                          */
+
+package org.scalajs.jsenv.phantomjs
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Promise, Future}
+import scala.util.control.NonFatal
+
+import java.io._
+import java.net._
+import java.nio.file.{Files, StandardCopyOption}
+
+import org.scalajs.jsenv._
+
+import org.scalajs.io._
+import org.scalajs.io.URIUtils.fixFileURI
+import org.scalajs.io.JSUtils.escapeJS
+
+private final class ComRun(jettyClassLoader: ClassLoader,
+    runConfig: RunConfig, onMessage: String => Unit,
+    startRun: VirtualBinaryFile => JSRun)
+    extends JSComRun {
+
+  import ComRun._
+
+  import runConfig.logger
+
+  private val promise = Promise[Unit]()
+
+  def future: Future[Unit] = promise.future
+
+  private val sendBuffer = mutable.ListBuffer.empty[String]
+  private var canSendNow: Boolean = false
```

Comment:
@gzm0 said:

> These two scream factory or Future.

(only applies to `canSendNow` at this point)

Reply:
I don't see how a `Future` can help, here. We can't just queue up messages to be sent as `onSuccess` handlers to the `Future`, because `Future` does not give any guarantee about the order in which the handlers are called, whereas our `send()` does.